### PR TITLE
[FSDP] Profiling range for FSDP.backward

### LINKED ISSUE #67714: Add some profiling ranges to FSDP
## 🚀 Feature

FSDP has many extra steps in forward, backward to all gather parameters, scatter gradients, run CPU -> GPU copies for CPU offload, etc. For easier to understand profiling, it would be useful to add ranges on some of these events so that we know the high-level context/operations of what some low level operators correspond to.

For example, the following stream sync in backwards by itself might be hard to diagnose - 
image

Although this sync is caused by synchronizing the current stream which we need to do when offloading gradients, so it would be good to label the code section which does this as so.

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

### PR DESCRIPTION
**Pull request by LinjianMa**
Add profiling for pre and post backward hooks, partially fixes #67714

### CODE REVIEW

**Review by rohan-varma — COMMENTED** [review]
Thank you for the PR, awesome work!

Although, this change profiles the process to _register_ the pre and post backward hooks, not the actual invocation of the hooks. I believe we'd like to have ranges for the actual execution of these hooks, not just registering them.

In order to do this, you'd want to add `record_function` similar to what you've already done, but inside of the pre-backward-hook (i.e. within `def _pre_backward_hook`). 

cc @zhaojuanmao

**Comment by LinjianMa** [comment]
> Thank you for the PR, awesome work!
> 
> Although, this change profiles the process to _register_ the pre and post backward hooks, not the actual invocation of the hooks. I believe we'd like to have ranges for the actual execution of these hooks, not just registering them.
> 
> In order to do this, you'd want to add `record_function` similar to what you've already done, but inside of the pre-backward-hook (i.e. within `def _pre_backward_hook`).
> 
> cc @zhaojuanmao

Ah you are right, thank you for pointing that out, just made a new commit accordingly

**Review by zhaojuanmao — COMMENTED** [review]
codes look good to me, will let @rohan-varma accept it. Before landing this PR, let us run some benchmarks to make sure no regression on performance, I will ping you about how to run benchmarks offline

**Comment by LinjianMa** [comment]
I believe the failed CI test is related to the backward compatibility of special Lagrangian polynomial based functions and is not because of this PR

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2611]: We might want to enter the record_function context after this. _pre_backward_hook can run many times (as it is a hook registered on tensors), but it only does actual work once, so let's profile just that.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2724]: This runs per FlatParameter so ideally there would be something in the string to indicate which (id? index? name?) parameter it is running for. But I can't think of a great way to get this information, so let's just save it for a future PR.

Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2724]: nit: "FullyShardedDataParallel._post_backward_hook"

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2608]: nit: "FullyShardedDataParallel._pre_backward_hook"

**Review by rohan-varma — COMMENTED** [review]
Thanks for the update! Have a couple more suggestions, after which it should be good to go (pending benchmark verification)

File: torch/distributed/fsdp/fully_sharded_data_parallel.py
Comment (LinjianMa) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py]: Done, thanks for the suggestion!

**Review by rohan-varma — APPROVED** [review]
Stamping, feel free to land once benchmarks are successful!

**Pull request closed without merging**
